src/file/format/ron.rs

```diff
@@ -19,56 +19,51 @@ pub fn parse(
-    Ok(match *value {
-        ron::Value::Option(ref value) => match value {
-            Some(value) => {
-                let value = from_ron_value(uri, value.as_ref())?;
-                Value::new(uri, value.kind)
-            }
-            None => Value::new(uri, ValueKind::Nil),
+    let kind = match value {
+        ron::Value::Option(value) => match value {
+            Some(value) => from_ron_value(uri, *value)?.kind,
+            None => ValueKind::Nil,
         },
 
-        ron::Value::Unit => Value::new(None, ValueKind::Nil),
+        ron::Value::Unit => ValueKind::Nil,
 
-        ron::Value::Bool(value) => Value::new(uri, ValueKind::Boolean(value)),
+        ron::Value::Bool(value) => ValueKind::Boolean(value),
 
-        ron::Value::Number(ref value) => {
-            if let Some(value) = value.as_i64() {
-                Value::new(uri, ValueKind::Integer(value))
-            } else if let Some(value) = value.as_f64() {
-                Value::new(uri, ValueKind::Float(value))
-            } else {
-                unreachable!();
-            }
-        }
-
-        ron::Value::Char(ref value) => Value::new(uri, ValueKind::String(value.to_string())),
+        ron::Value::Number(value) => match value {
+            ron::Number::Float(value) => ValueKind::Float(value.get()),
+            ron::Number::Integer(value) => ValueKind::Integer(value),
+        },
 
-        ron::Value::String(ref value) => Value::new(uri, ValueKind::String(value.clone())),
+        ron::Value::Char(value) => ValueKind::String(value.to_string()),
 
-        ron::Value::Seq(ref values) => {
-            let mut array = Vec::new();
+        ron::Value::String(value) => ValueKind::String(value),
 
-            for value in values.iter() {
-                array.push(from_ron_value(uri, value)?);
-            }
+        ron::Value::Seq(values) => {
+            let array = values
+                .into_iter()
+                .map(|value| from_ron_value(uri, value))
+                .collect::<Result<Vec<_>, _>>()?;
 
-            Value::new(uri, ValueKind::Array(array))
+            ValueKind::Array(array)
         }
 
-        ron::Value::Map(ref values) => {
-            let mut map = HashMap::new();
-
-            for (key, value) in values.iter() {
-                map.insert(
-                    key.clone().into_rust::<String>()?,
-                    from_ron_value(uri, value)?,
-                );
-            }
-
-            Value::new(uri, ValueKind::Table(map))
+        ron::Value::Map(values) => {
+            let map = values
+                .iter()
+                .map(|(key, value)| match key.clone().into_rust::<String>() {
+                    Ok(key) => match from_ron_value(uri, value.clone()) {
+                        Ok(value) => Ok((key, value)),
+                        Err(err) => Err(err),
+                    },
+                    Err(err) => Err(err.into()),
+                })
```

Comment:
How about

```rust
.into_iter() // use into_iter() here so you don't have to clone()
.map(|(key, value)| {
    let k = key.into_rust::<String>()?;
    let v = from_ron_value(uri, value)?;
    Ok((k, v))
})
```

Reply:
Unfortunately, `ron::Map` does not implement `into_iter`, so I can't do much about the cloning. But I have no idea why I used matches when I could've just let Rust handle the early returns for me. It's way too early in the morning.